bson/json_util.py

```diff
@@ -855,100 +856,188 @@ def _encode_int64(obj: Int64, json_options: JSONOptions) -> Any:
+
+
+def _encode_binary_obj(obj: Binary, json_options: JSONOptions) -> dict:
+    return _encode_binary(obj, obj.subtype, json_options)
+
+
+def _encode_uuid(obj: uuid.UUID, json_options: JSONOptions) -> dict:
+    if json_options.strict_uuid:
+        binval = Binary.from_uuid(obj, uuid_representation=json_options.uuid_representation)
+        return _encode_binary(binval, binval.subtype, json_options)
+    else:
+        return {"$uuid": obj.hex}
+
+
+def _encode_objectid(obj: ObjectId, dummy0: Any) -> dict:
+    return {"$oid": str(obj)}
+
+
+def _encode_timestamp(obj: Timestamp, dummy0: Any) -> dict:
+    return {"$timestamp": {"t": obj.time, "i": obj.inc}}
+
+
+def _encode_decimal128(obj: Timestamp, dummy0: Any) -> dict:
+    return {"$numberDecimal": str(obj)}
+
+
+def _encode_dbref(obj: DBRef, json_options: JSONOptions) -> dict:
+    return _json_convert(obj.as_doc(), json_options=json_options)
+
+
+def _encode_minkey(dummy0: Any, dummy1: Any) -> dict:
+    return {"$minKey": 1}
+
+
+def _encode_maxkey(dummy0: Any, dummy1: Any) -> dict:
+    return {"$maxKey": 1}
+
+
 # Encoders for BSON types
-_encoders = {
-    5: lambda obj, json_options: _encode_binary(obj, obj.subtype, json_options),  # Binary
-    7: lambda obj, json_options: {"$oid": str(obj)},  # noqa: ARG005 ObjectId
-    9: _encode_datetimems,  # DatetimeMS
-    13: _encode_code,  # Code
-    17: lambda obj, json_options: {"$timestamp": {"t": obj.time, "i": obj.inc}},  # noqa: ARG005 Timestamp
-    18: _encode_int64,  # Int64
-    19: lambda obj, json_options: {"$numberDecimal": str(obj)},  # noqa: ARG005 Decimal128
-    100: lambda obj, json_options: _json_convert(obj.as_doc(), json_options=json_options),  # DBRef
-    127: lambda obj, json_options: {"$maxKey": 1},  # noqa: ARG005 MaxKey
-    255: lambda obj, json_options: {"$minKey": 1},  # noqa: ARG005 MinKey
+# Each encoder function's signature is:
+#   - obj: a Python data type, e.g. a Python int for _encode_int
+#   - json_options: a JSONOptions
+_ENCODERS: dict[Any, Callable[[Any, JSONOptions], Any]] = {
+    bool: _encode_noop,
+    bytes: _encode_bytes,
+    datetime.datetime: _encode_datetime,
+    DatetimeMS: _encode_datetimems,
+    float: _encode_float,
+    int: _encode_int,
+    str: _encode_noop,
+    type(None): _encode_noop,
+    uuid.UUID: _encode_uuid,
+    Binary: _encode_binary_obj,
+    Int64: _encode_int64,
+    Code: _encode_code,
+    DBRef: _encode_dbref,
+    MaxKey: _encode_maxkey,
+    MinKey: _encode_minkey,
+    ObjectId: _encode_objectid,
+    Regex: _encode_regex,
+    RE_TYPE: _encode_regex,
+    Timestamp: _encode_timestamp,
+    Decimal128: _encode_decimal128,
+}
+
+_MARKERS: dict[Any, Any] = {
+    5: _encode_binary_obj,
+    7: _encode_objectid,
+    9: _encode_datetimems,
+    11: _encode_regex,
+    13: _encode_code,
+    17: _encode_timestamp,
+    18: _encode_int64,
+    19: _encode_decimal128,
+    100: _encode_dbref,
+    127: _encode_maxkey,
+    255: _encode_minkey,
```

Comment:
Out of scope for this PR, but do we have some constants/enum type object we could have these numbers reference for readability in the future? I imagine we need to handle these numerical types elsewhere, and I'm curious thoughts on that.

Reply:
Removed these magic constants by refactoring it to use `_type_marker`.